### ISSUE TITLE
dynamically "generating" properties

### ISSUE THREAD

**Issue by johannesschobel** [aedart/data-transfer-object-dto#2]
Hey there,

i am just playing around with your package - nice idea so far. However, i stumbled upon a question, i wasn't able to answer myself..

Is it possible to "dynamically" generate properties in a `DTO`? For example, i want to have one `GenericDTO` that automatically assigns `key/value` pairs..

However, if I call
```
$dto = new GenericDTO(['foo' => 'bar', ...]);`
```
i get an error, that `foo` is not defined or inaccessible.. Which is true, because my `GenericDTO` does not have such a property..

In other words, is it possible to `append` properties dynamically? If yes, how? ;)
Thanks a lot for your time / effort!

**Comment by aedart** [comment]
Hi @johannesschobel 
By default, you are not able to dynamically add properties to this version (interpretation) of a DTO.
That being said, of course you can create a modified version hereof, which enables you to do so.

You need to overwrite the `__set` method. The following is just an example (of the top of my head), so it might not work 100% - you should tinker around with this:

```php
    public function __set(string $name, $value) : void
    {
        $resolvedValue = $value;
        $methodName = $this->generateSetterName($name);
        if ($this->hasInternalMethod($methodName)) {
            $resolvedValue = $this->resolveValue($methodName, $value);
        }

        if ($this->hasInternalProperty($name)) {
            $this->invokeSetter($this->getInternalProperty($name), $value);
            return;
        }

        // Disable the exception if property does not exist
        //throw new UndefinedPropertyException(sprintf('Property "%s" does not exist or is inaccessible', $name));

        // Dynamically assign property to DTO
        $this->$name = $value;
    }
```

**Comment by johannesschobel** [comment]
Hey @aedart .. wow, thank you very much for your fast and detailled response to my question..

Do you think, it would be benefitial to add such a feature to this package? I am thinking about some kind of additional param you can pass in the constructor (and setter) called `strict = true`). This param, in turn, will allow to "dynamically generate" properties, if it is set to `false`.

Maybe, we can adjust your already existing `__set(...)` method and just check for this param? What do you think?

**Comment by aedart** [comment]
Hi again

Hmmm... well, if so, then I would most likely create another abstraction, e.g. `GenericDto` that can support this feature. It could inherit from the existing `DataTransferObject` and just overwrite the methods that it would require. Doing so will not break evt. existing usage and keep this kind of feature isolated.

If you are up for it, you are more than welcome to create a pull request for this. It is always nice to see what other developers can come up with.
Alternatively, you will have to wait until I can take some time off for this (X-mas, work, moving... lots of things going on in my life at the moment).

**Comment by aedart** [comment]
Hi @johannesschobel 
Since I haven't heard more on this matter, I will close this issue. Feel free to create another issue, if there is a need for it.
